packages/taste-test/src/TasteTest.class.ts

```diff
@@ -0,0 +1,234 @@
+
+    const userKey =
+      this.lucid.utils.getAddressDetails(walletAddress).paymentCredential?.hash;
+
+    if (!userKey) {
+      throw new Error("Missing wallet's payment credential hash.");
+    }
+
+    let coveringNode: UTxO | undefined;
+    if (args.utxos) {
+      coveringNode = args.utxos[0];
+    } else {
+      const nodeUTXOs = await this.lucid.utxosAt(nodeValidatorAddr);
+      coveringNode = findCoveringNode(nodeUTXOs, userKey);
+    }
+
+    if (!coveringNode || !coveringNode.datum) {
+      throw new Error("Could not find covering node.");
+    }
+
+    const coveringNodeDatum = Data.from(coveringNode.datum, SetNode);
+
+    const prevNodeDatum = Data.to(
+      {
+        key: coveringNodeDatum.key,
+        next: userKey,
+      },
+      SetNode
+    );
+
+    const nodeDatum = Data.to(
+      {
+        key: userKey,
+        next: coveringNodeDatum.next,
+      },
+      SetNode
+    );
+
+    const redeemerNodePolicy = Data.to(
+      {
+        PInsert: {
+          keyToInsert: userKey,
+          coveringNode: coveringNodeDatum,
+        },
+      },
+      DiscoveryNodeAction
+    );
+
+    const redeemerNodeValidator = Data.to("LinkedListAct", NodeValidatorAction);
+
+    const assets = {
+      [toUnit(nodePolicyId, `${fromText(SETNODE_PREFIX)}${userKey}`)]: 1n,
+    };
+
+    const correctAmount = BigInt(args.assetAmount.amount) + MIN_COMMITMENT_ADA;
+
+    const upperBound = (args?.currentTime ?? Date.now()) + TIME_TOLERANCE_MS;
+    const lowerBound = (args?.currentTime ?? Date.now()) - TIME_TOLERANCE_MS;
+
+    const tx = this.lucid
+      .newTx()
+      .collectFrom([coveringNode], redeemerNodeValidator)
+      .compose(this.lucid.newTx().attachSpendingValidator(nodeValidator))
+      .payToContract(
+        nodeValidatorAddr,
+        { inline: prevNodeDatum },
+        coveringNode.assets
+      )
+      .payToContract(
+        nodeValidatorAddr,
+        { inline: nodeDatum },
+        { ...assets, lovelace: correctAmount }
+      )
+      .addSignerKey(userKey)
+      .mintAssets(assets, redeemerNodePolicy)
+      .compose(this.lucid.newTx().attachMintingPolicy(nodePolicy))
+      .validFrom(lowerBound)
+      .validTo(upperBound);
+
+    return this.completeTx({ tx, referralFee: args.referralFee });
+  }
+
+  private async completeTx({
+    tx,
+    referralFee,
+  }: ITasteTestCompleteTxArgs): Promise<ITxBuilder<Tx, Datum | undefined>> {
+    if (referralFee) {
+      if (referralFee.payment.metadata.assetId !== "") {
+        tx.payToAddress(referralFee.destination, {
+          [referralFee.payment.metadata.assetId]: referralFee.payment.amount,
+        });
+      } else {
+        tx.payToAddress(referralFee.destination, {
+          lovelace: referralFee.payment.amount,
+        });
+      }
+    }
+
+    const baseFees: Omit<ITxBuilderFees, "cardanoTxFee"> = {
+      deposit: new AssetAmount(NODE_ADA, 6),
+      referral: referralFee?.payment,
+      foldFee: new AssetAmount(FOLDING_FEE_ADA, 6),
```

Comment:
What is the foldFee?

Reply:
once the taste test is done, someone has to walk over the list and aggregate up all the nodes, to produce the pool; the foldFee is the fee the user pays to cover that, currently set to 1 ADA.